Fix: add missing int64 <-> bigint primitive mapping when generating TypeScript definitions

Dabolus opened the pull request:
This PR adds a missing mapping between `int64_t` <=> `bigint` that makes the typings generation fail for code that makes use of that primitive type.

Fixes: https://github.com/emscripten-core/emscripten/issues/20450

brendandahl reviewed (changes requested):
Please add a test case to https://github.com/emscripten-core/emscripten/blob/main/test/other/embind_tsgen.cpp and update the reference file https://github.com/emscripten-core/emscripten/blob/main/test/other/embind_tsgen.d.ts

This may need to be a separate test since it needs bigint support.

Dabolus commented:
@brendandahl I added the mapping also for unsigned int64 and I updated the test file. I was able to test it locally and the output is as expected. I also updated the documentation page with the new mappings. Now I just have a couple of doubts:

1. I didn't need to create a new test file because the mapping works also if bigint support is not enabled. Is it correct or should we make the mapping only work when bigint support is enabled (using `#if WASM_BIGINT`)?
2. The types received by `typeToJsName` are `int64_t` and `uint64_t`, so everything works fine. However, in the documentation we have `long` and `unsigned long`. Should we call them `long long` and `unsigned long long` on the docs or should we leave them as they are? Note that I also tried using `long long` and `unsigned long long` in the code and the result is the same. `typeToJsName` will receive as input `int64_t` and `uint64_t` respectively.

sbc100 commented on site/source/docs/porting/connecting_cpp_and_javascript/embind.rst:
```diff
@@ -917,6 +917,10 @@ Out of the box, *embind* provides converters for many standard C++ types:
+| ``int64_t``         | BigInt                                                             |
++---------------------+--------------------------------------------------------------------+
+| ``uint64_t``        | BigInt                                                             |
++---------------------+--------------------------------------------------------------------+
```
Perhaps add `when building with `-sWASM_BIGINT`?

brendandahl commented:
> 1. I didn't need to create a new test file because the mapping works also if bigint support is not enabled. Is it correct or should we make the mapping only work when bigint support is enabled (using `#if WASM_BIGINT`)?

Yeah, we should really only enable when bigint is enabled since it won't work otherwise.

> 2. The types received by `typeToJsName` are `int64_t` and `uint64_t`, so everything works fine. However, in the documentation we have `long` and `unsigned long`. Should we call them `long long` and `unsigned long long` on the docs or should we leave them as they are? Note that I also tried using `long long` and `unsigned long long` in the code and the result is the same. `typeToJsName` will receive as input `int64_t` and `uint64_t` respectively.

I think what you have currently is clear. Like Sam mentioned, we should add a note or asterisk that this is only with bigint enabled.

Dabolus commented:
I initially made the change so that the generation would fail without the `-sWASM_BIGINT` flag, bug then I remembered this (from the FAQ):

> If you build using the _-sWASM_BIGINT_ flag, then _int64_t_ and _uint64_t_ will be represented as _bigint_ values in JS. Without the _-sWASM_BIGINT_ flag, the values will be represented as _number_ in JS which can't represent int64s, so what happens is that in exported functions (that you can call from JS) we "legalize" the types, by turning an i64 argument into two i32s (low and high bits), and an i64 return value becomes an i32, and you can access the high bits by calling a helper function called getTempRet0.

So now I made it generate `bigint`s if their support is enabled, `number`s otherwise. I also added the related test case and addressed it in the documentation, let me know if it makes sense.

Dabolus commented:
I see that my new test [passed on CI](https://app.circleci.com/pipelines/github/emscripten-core/emscripten/31316/workflows/004d98e1-124f-4a54-b419-7c5161411549/jobs/700589?invite=true#step-110-134851_58), but other unrelated tests failed, not sure why 🤔

brendandahl commented on src/embind/embind_ts.js:
```diff
@@ -210,6 +210,13 @@ var LibraryEmbind = {
+#if WASM_BIGINT
+        ['int64_t', 'bigint'],
+        ['uint64_t', 'bigint'],
+#else
```
IIRC, if bigint is not enabled, won't using a `int64_t` cause a runtime type error? I'm thinking we should not emit a map for these values if they'll be broken anyway.

Dabolus replied:
@brendandahl not sure if I'm getting something wrong, but according to the FAQ when the `WASM_BIGINT` flag is not provided the int64 values will be legalized to standard 32 bit numbers, and the missing higher bits can be accessed by using the `getTempRet0` helper

Dabolus replied:
I updated the code so that the mapping is only enabled when `WASM_BIGINT` flag is present, failing otherwise. I also updated the test and the docs. Let me know if everything looks good now.

brendandahl commented:
> I see that my new test [passed on CI](https://app.circleci.com/pipelines/github/emscripten-core/emscripten/31316/workflows/004d98e1-124f-4a54-b419-7c5161411549/jobs/700589?invite=true#step-110-134851_58), but other unrelated tests failed, not sure why 🤔

We have some flaky tests, so probably unrelated.
